Adds London air component

robmarkcole opened the pull request:
## Description:
The london_air component queries the London air quality data feed provided by Kings College London. A single sensor will be added for each location (local authority district or borough) specified in the configuration file. The state of each sensor is the overall air quality in that borough.

Replaces [8813](https://github.com/home-assistant/home-assistant/pull/8813)
**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io/pull/3132) <3132>

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: london_air
    locations:
      - Barking and Dagenham
      - Bexley
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

If the code does not interact with devices:
  - [x ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x ] Tests have been added to verify that the new code works.

robmarkcole committed:
Adds London air component

robmarkcole committed:
Fix lints

fabaff commented on homeassistant/components/sensor/london_air.py:
```diff
@@ -0,0 +1,210 @@
+        if sites_status:
+            self._state = max(set(sites_status), key=sites_status.count)
+        else:
+            self._state = 'no_species_data'
```
This looks like the equivalent to `STATE_UNKNOWN`.

robmarkcole replied:
Yes, using STATE_UNKNOWN now

fabaff commented on homeassistant/components/sensor/london_air.py:
```diff
@@ -0,0 +1,210 @@
+    @Throttle(SCAN_INTERVAL)
+    def update(self):
+        """Get the latest data from TFL."""
+        response = requests.get(URL)
```
Please add a timeout to the request.

robmarkcole replied:
Done

balloob committed:
Polymer 2 (#8815)

* Update build for Polymer 2

* Update webcomponents polyfills/helpers

* Load ES5 class adapter when not in dev mode

* Update frontend

robmarkcole committed:
Fix variable

fabaff commented on homeassistant/components/sensor/london_air.py:
```diff
@@ -0,0 +1,210 @@
+            self._state = 'no_species_data'
+
+
+def parse_api_response(response):
```
Please consider to create a Python module which will become a requirement for this sensor. This way other users (outside of Home Assistant) can profit from your work with the parsing of the data.

robmarkcole replied:
Good idea

robmarkcole replied:
I think I will create a more fully fledged module for external use at a later date

JulianKahnert committed:
geizhals sensor component (#8458)

* initial create of the geizhals component

* only .coveragerc, geizhals.py, and requirements_all.txt included

azogue committed:
Xiaomi vacuum as platform of new `vacuum` component derived from ToggleEntity, and services (#8623)

* Xiaomi vacuum as component with switch, sensors and services

- Conversion from switch platform to async component.
- Add services proposed in #8416 to the new component, with shorter names.
- Add sensors for the vacuum robot as a selectable list from `battery`, `state`, `error`, `fanspeed`, `clean_time` and `clean_area` (the state attributes of the switch). The sensors don't poll, but listen to a signal to update the state, the switch fires this signal when updating.
- Assign default icons to sensors and the switch (`mdi:google-circles-group` looks like the robot!)

* path change in requirements_all (from switch platform to component)

* copy pasting is a bad habit

* services to the components services.yaml, modify .coveragerc

* review: use with multiple hosts, fix calls to async_add_devices, fix ranges for services

* `icon_for_battery_level` util method

* Xiaomi vacuum as platform of new component vacuum

- Created new component `vacuum` from a ToggleEntity.
- Add services `turn_on`, `turn_off`, `cleaning_play_pause`, `stop`, `return_to_base`, `locate`, `set_fanspeed` and `send_command`.
- Remove the main switch for the xiaomi vacuum (the toggable main entity is the switch).
- Add `support flags` for the common services
- Assign default icons to sensors and the switch (`mdi:google-circles-group` looks like the robot!)
- Move services descriptions to a yaml file for the new component.
- Update requirements_all.
- Update coveragerc.

* fix coveragerc

* fix battery icon helper to use more icons

* remove sensors, create properties and support flags for custom UI

* cleaning

* updated state_attrs for filtering in UI, renamed platform to simply `xiaomi`

* fix platform rename

* change fanspeed and expose `fanspeed_list` to use speed steps

* minor fixes

- Rename service `start_pause`
- Add 'Error' attribute only if `got_error`.
- Minor changes

* rename state attrs

* rename state attrs

* review changes: cut fan__speed, style changes, remove logging, and more

* add ATTR_COMMAND = 'command' to const

* pop entity_id from service data

* remove property accessor for vacuum object

* lint fix

* fix extra attrs names

* module level functions for calling the services

* params as optional keyword for `send_command`

* params as optional keyword for `send_command`, remove debug logs

* explicit parameters for `set_fan_speed` and `send_command`

* Demo platform for the vacuum component

* vacuum tests for the Demo platform

* some fixes

* don't omit vacuum

* vacuum tests for the Xiaomi platform

* fix test

* fix

* fix xiaomi test

* fix coveragerc

* test send command

* fix coveragerc

* fix string formatting

* The coverage is to low. It need 93% or more

fanthos committed:
Update yweather.py (#8820)

Fix missing weather unit support.
Move some weather code to their correct classes.

Hellowlol committed:
Make HA discover sabnzbd and add it to the Configurator (#8634)

* Init discover sab.

* Fix hound errors

Nobody likes being hound at :(

* sabnzbd discovery says if ssl is active.

* Fixups after codereview.

CharlesBlonde committed:
Fix Dyson sensors if devices are configured without standby monitoring. Fixes #8569 (#8826)

Upgrade libpurecoolink libraries without unused enum34 dependency

balloob committed:
Update frontend

balloob committed:
Update frontend 2

balloob committed:
Block dependencies that depend on enum34 (#8698)

* Block dependencies that depend on enum34

* Remove uninstalling enum34

* Update validation script

* Add constraints to tox.ini

* Upgrade yeelight to version that uses enum-compat

* Disable sensor.skybeacon

* Lint

fabaff committed:
Upgrade aiohttp to 2.2.5 (#8828)

OverloadUT committed:
History query and schema optimizations for huge performance boost (#8748)

* Add DEBUG-level log for db row to native object conversion

This is now the bottleneck (by a large margin) for big history queries, so I'm leaving this log feature in to help diagnose users with a slow history page

* Rewrite of the "first synthetic datapoint" query for multiple entities

The old method was written in a manner that prevented an index from being used in the inner-most GROUP BY statement, causing massive performance issues especially when querying for a large time period.

The new query does have one material change that will cause it to return different results than before: instead of using max(state_id) to get the latest entry, we now get the max(last_updated). This is more appropriate (primary key should not be assumed to be in order of event firing) and allows an index to be used on the inner-most query. I added another JOIN layer to account for cases where there are two entries on the exact same `last_created` for a given entity. In this case we do use `state_id` as a tiebreaker.

For performance reasons the domain filters were also moved to the outermost query, as it's way more efficient to do it there than on the innermost query as before (due to indexing with GROUP BY problems)

The result is a query that only needs to do a filesort on the final result set, which will only be as many rows as there are entities.

* Remove the ORDER BY entity_id when fetching states, and add logging

Having this ORDER BY in the query prevents it from using an index due to the range filter, so it has been removed.

We already do a `groupby` in the `states_to_json` method which accomplishes exactly what the ORDER BY in the query was trying to do anyway, so this change causes no functional difference.

Also added DEBUG-level logging to allow diagnosing a user's slow history page.

* Add DEBUG-level logging for the synthetic-first-datapoint query

For diagnosing a user's slow history page

* Missed a couple instances of `created` that should be `last_updated`

* Remove `entity_id` sorting from state_changes; match significant_update

This is the same change as 09b3498f410106881fc5e095c49a8d527fa89644 , but applied to the `state_changes_during_period` method which I missed before. This should give the same performance boost to the history sensor component!

* Bugfix in History query used for History Sensor

The date filter was using a different column for the upper and lower bounds. It would work, but it would be slow!

* Update Recorder purge script to use more appropriate columns

Two reasons: 1. the `created` column's meaning is fairly arbitrary and does not represent when an event or state change actually ocurred. It seems more correct to purge based on the event date than the time the database row was written.
2. The new columns are indexed, which will speed up this purge script by orders of magnitude

* Updating db model to match new query optimizations

A few things here: 1. New schema version with a new index and several removed indexes
2. A new method in the migration script to drop old indexes
3. Added an INFO-level log message when a new index will be added, as this can take quite some time on a Raspberry Pi

azogue committed:
Update xiaomi vacuum tests and include in coverage (#8845)

* Fix tests for Demo vacuum platform (and increase coverage)

* increase coverage of xiaomi vacuum tests and include in coverage

Also little fixes

* remove print statement

jeroenterheerdt committed:
Egardia (#8389)

* Added support for Egardia / Woonveilig alarm control panel

* Added support for Egardia / Woonveilig alarm control panel

* Added support for Egardia / Woonveilig alarms

* Updating egardia support with exception handling and other fixes

* Egardia platform, requirements file updated

* Fixing state checking

* Adding exception handling

* Removing unnecessary logging

* Removing unnecessary logging

* Updating to egardiadevice component 1.0.10

* Improving exception handling

* Adding implementation of egardiaserver for alarm triggered status

* Clean-up

* Fix my previous change

robmarkcole committed:
Changes requested

Address changes requested by @fabaff

fabaff committed:
Do not call update() in constructor. (#8837)

fabaff committed:
Upgrade sendgrid to 4.2.1 (#8839)

fabaff committed:
Do not call update() in constructor (#8840)

fabaff committed:
Make 'monitored_conditions' optional (#8843)

* Do not call update() in constructor

* Make 'monitored_conditions' optional

* Update tests

fabaff committed:
Do not call update() in constructor (#8849)

* Do not call update() in constructor

* Fix pylint issues

fabaff committed:
Make 'monitored_conditions' optional (#8848)

* Do not call update() in constructor

* Update tests

fabaff committed:
Do not call update() in constructor (#8847)

fabaff committed:
Catch ConnectionRefusedError (#8844)

* Do not call update() in constructor

* Catch ConnectionRefusedError

fabaff committed:
Upgrade sqlalchemy to 1.1.13 (#8850)

fabaff committed:
Upgrade pylast to 1.9.0 (#8854)

pschmitt committed:
Wi-Fi enabled Roomba support (#8825)

* Roomba vacuum component

* Update requirements and coveragerc

* Update error handling message

* Implement changes requested by @azogue

* Add missing import

* Don't wrap commands with functools.partial

* Refactoring

* Remove state attribute and use double quotes for log messages strings

* Remove unused constants

* Sorting

* Sorting + remove None arg from dict.get() calls

* Re-sort imports

CharlesBlonde committed:
Add support to Dyson 360 Eye robot vacuum using new vacuum platform (#8852)

* Add support to Dyson 360 Eye robot vacuum using new vacuum platform

* Fix tests with Python 3.5

* Code review

* Code review - v2

* Code review - v3

robmarkcole commented:
@fabaff I don't understand the failed test - everything passed before my last commit and is passing on my machine

fabaff committed:
Upgrade pyasn1 to 0.3.2 and pyasn1-modules to 0.0.11 (#8856)

fabaff committed:
Allow to set coordinates (#8858)

fronzbot committed:
Added rounding to Google Wifi (#8866)

turbokongen committed:
Fix off_delay for zwave trigger sensors (#8864)

foxel committed:
Do not use pychromecast.Chromecast for Cast Groups (#8786)

* Do not use pychromecast.Chromecast for Cast Groups

pychromecast.Chromecast creates Chromecast instance with friendly_name and cast_type of the device and not of a group.
Which leads to collisions

* Update cast.py

* using hass.data

* Fixed and extended tests

* Line length in tests

* Lint in tests

balloob committed:
Allow get local ip to work without internet (#8855)

fabaff committed:
Do not call update() in constructor (#8859)

azogue committed:
Add new service `clean_spot` to vacuums (#8862)

* Add new service `clean_spot` to vacuums

    - Add as base component service, with associated support flag to make it optional
    - Implement on Demo vacuum
    - Implement on Xiaomi vacuum
    - Update tests for platforms Demo and Xiaomi
    - Change default icon for vacuums to `mdi:roomba`, but keep the one for the Xiaomi
    - (In a polymer PR: add new service to command toolbar in the 'more-info' card)

* Add `clean_spot` service description

* fix default properties for vacuum component

balloob committed:
Warn instead of raise on duplicate YAML key (#8834)

* Warn instead of raise on duplicate key

* Update test_yaml.py

* Lint

* Change to error

PhracturedBlue committed:
Add Initial Mailbox panel and sensor (#8233)

* Initial implementation of Asterisk Mailbox

* Rework asterisk_mbox handler to avoid using the hass.data hash.  Fix requirements.

* Handle potential asterisk server disconnect.  bump asterisk_mbox requirement to 0.4.0

* Use async method for mp3 fetch from server

* Add http as dependency

* Minor log fix. try to force Travis to rebuild

* Updates based on review

* Fix error handling as per review

* Fix error handling as per review

* Refactor voicemail into mailbox component

* Hide mailbox component from front page

* Add demo for mailbox

* Add tests for mailbox

* Remove asterisk_mbox sensor and replace with a generic mailbox sensor

* Fix linting errors

* Remove mailbox sensor.  Remove demo.mp3.  Split entity from platform object.

* Update mailbox test

* Update mailbox test

* Use events to indicate state change rather than entity last-updated

* Make mailbox platform calls async.  Fix other review concerns

* Rewrite mailbox tests to live at root level and be async.  Fixmailbox dependency on http

* Only store number of messages not content in mailbox entity

tlyakhov committed:
Add Leviton Decora Smart WiFi Device Platform (#8529)

* Add Leviton Decora Smart WiFi Device Platform

* Decora WiFi Code Review Fixes

balloob committed:
Update mailbox panel icon

balloob committed:
Consolidate config panels (#8857)

* Remove automation panel registration

* Move Z-Wave config API to config.zwave

* Remove no longer needed test

* Lint

* Update frontend

Klikini committed:
Fix spelling error and update link (#8869)

pschmitt committed:
Implement Roomba fan speed (#8863)

* Implement Roomba fanspeed

* Fix: fan_speed_list is always empty

* Log instead of raising an exception when incorrect fan speed has been provided

* Don't attempt to set any preference if fan speed is invalid

pschmitt committed:
Add nuki lock'n'go and unlatch services and add attributes (#8687)

* Add lock'n'go service

* Add unlatch service

* Implement changes requested by @MartinHjelmare

* Fix service domain

leppa committed:
RFLink: Add send_command service (#8876)

Add an optional extended description…

fabaff committed:
Do not call update() in constructor (#8881)

bachya committed:
Add RainMachine switch platform (#8827)

* Add RainMachine switch platform

* Updated requirements_all.txt

* Cleaning up CI and coverage results

* Small update to deal with older pylint

* Fixed small indentation-based error

* Added some more defensive try/except logic around calls

* I'm not a fan of importing a library multiple times :)

* Making PR-requested changes

* Fixed ref to positional parameter

* Attempting to fix broken linting

* Ignoring no-value-for-parameter pylint error

fabaff committed:
Allow usage of colorlog 3.0.1 (#8885)

fabaff committed:
Change level (#8883)

fabaff committed:
Upgrade youtube_dl to 2017.8.6 (#8880)

spirit-x committed:
Fix media_extractor for some sites (#8887)

fabaff committed:
Do not call update() in constructor (#8878)

* Do not call update() in constructor

* Fix lint issues

fabaff committed:
Do not call update() in constructor (#8892)

bachya committed:
Changed Pi-hole graphs from stacked bar to line (#8896)

PhracturedBlue committed:
Add longer text strings to mailbox demo to test string truncation (#8893)

* Add longer text strings to mailbox demo to test string truncation in frontend

* Remove lorem ipsum txt file

* Use format instead of %

kfcook committed:
added support for setting/getting position of lutron caseta covers (#8898)

balloob committed:
Update frontend

karlkar committed:
Fix for Neato D3 Connected state obtaining (#8817)

Commit:
Update simplisafe-python version (#8908)

amelchio committed:
LIFX: improve performance of multi-light transitions (#8873)

* LIFX: improve performance of multi-light transitions

To avoid hub overload, the light.turn_on call will change each light
sequentially.

As LIFX has no hub we can safely increase performance by starting all
light transitions concurrently.

* Improve state updates after light changes

The light.turn_on call will set a new state and then immediately read it
back. However, reading the state of a LIFX light right after a state
change can still return the old value.

To handle this situation we have previously delayed the update request a
little while to allow a potential state change to settle. Because light
updates are now run in parallel, this delay might be too short when many
lights are set at once.

This commit introduces a per-light Lock to make it explicit when the
state cannot yet be trusted.

We must then do the state update ourselves. This was already done at the
end of a long transition and that code can be reused for also doing the
update at the start of a transition.

molobrakos committed:
Support media position and media duration (will display progressbar in ui) (#8904)

abmantis committed:
Add new device tracker for Huawei Routers. (#8488)

* Add new device tracker for Huawei Routers.

	This was tested with the HG8247H model, used by Vodafone
	Portugal for the Fiber service.

* add to .coveragerc; remove import and space

* add comments and fix lint

* rename methods

* huawei_router: add constants to scanner class

* huawei_router: remove lock; use format() in string

* huawei_router: use tupple instead of member only class

* huawei_router: reduce min scan time

* huawei_router: lint

* huawei_router: lint

* huawei_router: add missing lines in imports

* huawei_router: correctly decode string after router firmware update

* Remove things that is done on core now

fabaff committed:
Add Shodan sensor (#8902)

mar-schmidt committed:
Added possibilities to use template in the command_line sensor (#8505)

* Added possibilities to use template in the command_line sensor

* Minor style guideline conforms

* Minor style guideline conforms

* Added new test for template rendering

* Minor style guideline conforms

* Minor style guideline conforms

* Fixed failing testcases

* Fix style violations

* fix code pretty

fabaff committed:
Honor PEP8 naming convention (#8909)

* Honor PEP8 naming convention

* Update validator

fabaff committed:
Version bump to 0.52.0.dev0

Commit:
Update python-wink version to fix Dome water valve bug. (#8923)

pschmitt committed:
Update roombapy to 1.3.1 to avoid installing all the mapping dependencies (#8925)

MartinHjelmare committed:
Remove not needed call to update (#8930)

* This will ensure no I/O in entity properties.

fabaff committed:
Add version sensor (#8912)

* Add version sensor

* Set version directly

* Rework tests and fix typo

* Remove additional blank line

robmarkcole reviewed (commented):
Have addressed these

groth-its committed:
Fix hue lights for Philips and non-philips lights (#8905)

cribbstechnologies committed:
fixing emulated hue issue and testing it (#8928)

* fixing emulated hue issue and testing it

* fixing hound issues

* I should probably stop using vim

* Check against dict directly instead of items.

MartinHjelmare committed:
Fix SET_TEMPERATURE_SCHEMA in climate component (#8879)

* Require either temperature or high/low target temperatures.
* Add tests.

balloob committed:
Update frontend

arsaboo committed:
Fixed cert_expiry sensor to delay firing on HA startup (#8920)

* Fixed cert_expiry sensor to delay firing on HA startup

* Addressed Travis complaints

* Added imports

* Fixed cert_expiry sensor to delay firing on HA startup

* Changed comment

syssi committed:
Remove spaces from Xiami switch attributes (#8952)

* Attributes of the xiaomi zigbee plug changed.

* Reformat.

azogue committed:
fix DeviceException handling when updating xiaomi vacuum (#8954)

* Fix DeviceException handling when updating entity

* add DeviceException error handling to generic request

schmittx committed:
Add service to alarm control panel for night mode arming (#8614)

* Update const.py

* Update __init__.py

* Update services.yaml

* Update totalconnect.py

* Update manual.py

Add night arm service for manual alarm control panel

* Update test_manual.py

Add tests for night mode arming

* Update manual.py

Fix docstring

MartinHjelmare committed:
Fix call to ha_send_commands (#8956)

* Name keyword arguments correctly according to dependency lib.
* Only pass keyword arguments that are not None.

balloob committed:
Sabnzbd: do not assume discovery info is a dict (#8951)

balloob committed:
fix issue #8948 in pushbullet (#8965)

* fix issue #8948 in pushbullet

* pushbullet

andrey-git committed:
Fix zwave power_consumption attribute (#8968)

andrey-git committed:
Turn foscam verbose mode off (#8967)

balloob committed:
Update frontend

emlove committed:
Add support for Automatic OAuth2 authentication (#8962)

* Add support for Automatic OAuth2 authentication

* Fix async conversion of configurator

* Rename method for async

* Use hass.components to get configurator component

* Fix typo

* Move session data to hidden directory

* Make configurator callback optional

pschmitt committed:
Nello.io lock support (#8957)

* Initial Nello.io lock support

* Log an error when unlocking failed

* Make the lock's state always locked

iamjackg committed:
Added continue-on-errors, added value template (#8971)

* Added continue-on-errors, added value template

* Refactored long lines

* Fixed whitespace issues

bachya commented:
@robmarkcole This line is what's failing in your CI:

```
R:150, 4: Too many nested blocks (6/5) (too-many-nested-blocks)
```

Danielhiversen committed:
Is_allowed_path raise for None path (#8953)

* is_allowed_path

* Fix #8948

* assert path is not None

* Update test_core.py

* Update test_core.py

* Update test_core.py

tlyakhov committed:
Fix #8960 - Decora Wi-Fi Switch unable to set brightness (#8989)

bachya committed:
Adds port/SSL config options for RainMachine (#8986)

* Adding port/SSL config updates

* New requirements generated

* Made `port` and `ssl` parameters optional

* Add defaults for new parameters

* Re-adding guard clause

* pass > continue

emlove committed:
Automatic device tracker remove password (#9002)

* Remove now disabled password auth from automatic

* Fallback to configurator more permissively

* Fix test for changes

* Bump lib

balloob committed:
Add scripts editor backend (#8993)

* Add scripts editor backend

* Fix docstrings

robmarkcole commented:
I need to rewrite parse_api_response()

balloob committed:
Update frontend

Commit:
MQTT Switch - Add configurable availability payload (#8934)

* Add configurable availabilty payload

* Fix

* Fix

* Lint fixes

* Fix tests

* Fix tests

* Move from const.py to mqtt switch

* New test

* Fix flake*

mjj4791 committed:
Buienradar newconditions (#8897)

* new monitored conditions and support for new weathercard

* new monitored conditions and support for new weathercard

* minor changes

karlkar committed:
Pushbullet, fix multiple messages sent when url param is set (#9006)

Cinntax committed:
Update to pyenvisalink 2.2, and remove range validation on zonedump i… (#8981)

* Update to pyenvisalink 2.2, and remove range validation on zonedump interval.

* Keep using default timer dump variable, only remove minimum check.

* Fix lint issue

* Indentation issue

pvizeli commented:
Yeah that look very ugly

hertg committed:
Fix #9010 - Swiss Public Transportation shows departure time in the past (#9011)

BioSehnsucht committed:
Add HipChat notify service. (#8918)

* Add HipChat notify service.

* Change HipChat notify service to use python-simple-hipchat-v2.

* Change HipChat notify service to use hipnotify

* Change HipChat notify service to remove redundant validation

especially-relative committed:
Update onkyo-eiscp to 1.2.3 (#9019)

robmarkcole committed:
Refactor parse_api_response()

homeassistant commented:
Hello @robmarkcole,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`2e9768749e23e8fe142138e3e26b66d0e014c00b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;8814&#x2F;commits&#x2F;2e9768749e23e8fe142138e3e26b66d0e014c00b) - This commit has something that looks like an email address (spirit-x@users.noreply.github.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

robmarkcole committed:
Merge remote-tracking branch 'home-assistant/dev' into london_air

homeassistant commented:
Hello @robmarkcole,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`2e9768749e23e8fe142138e3e26b66d0e014c00b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;8814&#x2F;commits&#x2F;2e9768749e23e8fe142138e3e26b66d0e014c00b) - This commit has something that looks like an email address (spirit-x@users.noreply.github.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:
